src/Compilers/Core/Portable/Text/ChangedText.cs

```diff
@@ -83,8 +83,16 @@ public override void CopyTo(int sourceIndex, char[] destination, int destination
         public override SourceText WithChanges(IEnumerable<TextChange> changes)
         {
             // compute changes against newText to avoid capturing strong references to this ChangedText instance.
-            var changed = (ChangedText)_newText.WithChanges(changes);
-            return new ChangedText(this, changed._newText, changed._changes);
+            // note: _newText is never a ChangeText, so if WithChanges is a no-op the result will not be a ChangedText
+            var changed = _newText.WithChanges(changes) as ChangedText;  
+            if (changed != null)
+            {
+                return new ChangedText(this, changed._newText, changed._changes);
+            }
+            else 
+            {
+                return this;
```

Comment:
@Pilchie okay, I've change the comment to be clearer and added assert when ChangedText is constructed to check that _newText is one of CompositeText, SubText, StringText and LargeText which are the only types used to represent the new text in the default SourceText.WithChanges implementation, which is the only implementation (unless we add a new one) that can even construct a ChangedText instance.